lib/services/access_list.go

```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+	isOwner := slices.ContainsFunc(accessList.GetOwners(), func(owner accesslist.Owner) bool {
 		return owner.Name == identity.Username
 	})
-	if ownerIdx == -1 {
-		return accessDenied
+	if !isOwner {
+		err := recursiveIsAccessListOwnerCheck(ctx, members, identity, accessList)
+		if err != nil {
+			return false, accessDenied
+		}
+		inherited = true
 	}
 
 	// Does the supplied Identity meet the ownership requirements?
 	if !UserMeetsRequirements(identity, accessList.Spec.OwnershipRequires) {
-		return accessDenied
+		return inherited, accessDenied
 	}
 
 	// We've gotten through all the checks, so the user is an owner.
-	return nil
+	return inherited, nil
 }
 
 // AccessListMembershipChecker will check if users are members of an access list and
 // makes sure the user is not locked and meets membership requirements.
 type AccessListMembershipChecker struct {
-	members AccessListMemberGetter
+	members AccessListsAndMembersGetter
 	locks   LockGetter
 	clock   clockwork.Clock
 }
 
 // NewAccessListMembershipChecker will create a new access list membership checker.
-func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListMemberGetter, locks LockGetter) *AccessListMembershipChecker {
+func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListsAndMembersGetter, locks LockGetter) *AccessListMembershipChecker {
 	return &AccessListMembershipChecker{
 		members: members,
 		locks:   locks,
 		clock:   clock,
 	}
 }
 
-// IsAccessListMember will return true if the user is a member for the current list.
-func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+// getAccessListDynamicMembers returns a list of all dynamic members (nested lists) for a given Access List.
+func getAccessListDynamicMembers(ctx context.Context, clt AccessListsAndMembersGetter, entry string) ([]string, error) {
+	var pageToken string
+	var dynamicMembers []string
+
+	for {
+		var members []*accesslist.AccessListMember
+		var err error
+
+		members, pageToken, err = clt.ListAccessListMembers(ctx, entry, 0 /* default page size */, pageToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, member := range members {
+			if member.Spec.MembershipKind == accesslist.MembershipKindList {
+				dynamicMembers = append(dynamicMembers, member.GetName())
+			}
+		}
+
+		if pageToken == "" {
+			break
+		}
+	}
+
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+	seen := map[string]struct{}{}
+	type queueItem struct {
+		name  string
+		depth int
+	}
+	queue := []queueItem{{name: accessList.GetName(), depth: 0}}
+	membershipErr := trace.NotFound("user %s is not a member of the access list or its parents", identity.Username)
+
+	for len(queue) > 0 {
```

Comment:
TBH I find the logic both here and in `recursiveIsAccessListOwnerCheck` (which is basically the same but for owners) slightly confusing. Similar to my comment below in `lib/services/local/access_list.go` - all this logic basically comes down to expanding either full owners or full members list, and then checking if the current identity is in there (plus the requirements check).

So with that in mind, can we have two helper methods: `ExpandAccessListOwners` and `ExpandAccessListMembers` that return full set of owners/members respectively? I provided an example of what an expand method would look like in that comment. Then you can use those helpers here and in the local service as well, and this expansion logic will be well-encapsulated and would lead to a more straightforward logic and probably better code reuse as well.

WDYT, can you refactor it like this?

Reply:
I could give that a shot. By 'full set of members/owners', do you mean ignoring any membership/ownership requirements, and checking those separately?

Reply:
If so, a refactor to `ExpandAccessListMembers` could work, but Owners don't have a reference to the Access List they're an owner of, so it might be more complicated to split out fetching them from verifying their ownership.